index.js

```diff
@@ -10,125 +10,127 @@ import {
-  });
+  return;
 };
 
-const debounce = (handleSearch, delay) => {
-  let timer;
+const todoStateFilter = (status = ALL, todos) => {
+  if (status === ALL) {
+    return todos;
+  } else if (status === COMPLETE) {
+    return todos.filter((todo) => todo.isDone === true);
+  }
+  return todos.filter((todo) => todo.isDone === false);
+};
 
-  return function (...args) {
-    clearTimeout(timer);
+const searchFilter = (searchText, todos) => {
+  if (searchText !== "") {
+    return todos.filter((task) =>
+      task.value.toLowerCase().includes(searchText)
+    );
+  }
+  return todos;
+};
 
-    timer = setTimeout(() => {
-      handleSearch(...args);
-    }, delay);
+const debounce = (searchFilter, delay) => {
+  let timer;
+  return function () {
+    clearTimeout(timer);
+    return new Promise((resolve) => {
+      timer = setTimeout(() => {
+        const result = searchFilter.apply(this, arguments);
+        resolve(result);
+      }, delay);
+    });
   };
 };
 
-const handleSearch = () => {
-  keepCard1Elements();
-
-  const searchText = searchInput.value.toLowerCase().trim();
+if (todos.length === 0) {
+  disableFilterBar();
+}
 
-  filteredTasks = filteredTodos?.length
-    ? filteredTodos.filter((task) =>
-        task.value.toLowerCase().includes(searchText)
-      )
-    : todos.filter((task) => task.value.toLowerCase().includes(searchText));
+export const renderTodoList = (todos) => {
+  let auxilaryTodos = [];
+  auxilaryTodos = [...searchFilter(searchText, todos || [])];
+  auxilaryTodos = [...todoStateFilter(filterStatus, auxilaryTodos)];
+  auxilaryTodos = [...paginationFilter(auxilaryTodos)];
+  taskList.innerHTML = "";
+  taskList.appendChild(taskInputCard);
 
-  filteredTasks.map((task) => {
-    const newTask = createTaskElement(task);
-    taskList.insertBefore(newTask, taskList.children[1]);
+  auxilaryTodos.map((todo) => {
+    const newTask = createTaskElement(todo);
+    taskList.append(newTask);
   });
 };
 
-export const renderTodoList = () => {
-  !todos.length
-    ? (initialTaskContainer.classList.remove("hide"),
-      initialTaskContainer.classList.add("show"),
-      (taskInputCard.style.display = "none"))
-    : (initialTaskContainer.classList.remove("show"),
-      initialTaskContainer.classList.add("hide"));
-
-  keepCard1Elements();
-  filteredTasks = [];
-  searchInput.value = "";
-
-  todos.length
-    ? tabs.forEach((tab) => {
-        tab.addEventListener("click", handleTabClick);
-        tab.removeAttribute("disabled");
-        tab.classList.remove("selected");
-        if (tab.id === "all") {
-          tab.classList.add("selected");
-        }
-      })
-    : tabs.forEach((tab) => {
-        tab.setAttribute("disabled", true);
-        tab.classList.remove("selected");
-      });
-
-  todos.map((todo) => {
-    const newTask = createTaskElement(todo);
-    taskList.insertBefore(newTask, taskList.children[1]);
-  });
+export const filterTasksByStatus = (status) => {
```

Comment:
let's move this functions to a new file. e.g: `actions.js`
Other functions such as, `todoStateFilter`, `searchFilter` should also be moved to this file.